diesel/src/sqlite/connection/from_sqlite_value.rs

```diff
@@ -0,0 +1,83 @@
+use super::ffi;
+
+/// The sqlite3_value_*type*-functions are only safe for so-called "protected" values.
+///
+/// Per https://www.sqlite.org/c3ref/value.html, values sent to application-defined functions are protected,
+/// making it safe to use those values with FromSqliteValue.
+pub trait FromSqliteValue {
```

Comment:
Why do we need this trait? It's entirely redundant with `FromSql`

Reply:
The `sqlite3_create_function_v2` gives us `sqlite3_value` pointers to work with. As far as I can tell, the only way to get the actual values out of these pointers is with the `sqlite3_value_*` functions. `FromSql` seems to use (via `SqliteValue`) the functions named `sqlite3_column_*`.

It might very well be that I have misunderstood something, but it seems to me that we need this in addition to `FromSql`.

Reply:
Can we change `SqliteValue` to operate on `sqlite3_value_*` functions, and construct it by calling `sqlite3_column_value`?

Reply:
Barring any hidden dragons, I think it would be possible. It would, however, require an extra copy, via [`sqlite3_value_dup`](https://www.sqlite.org/c3ref/value_dup.html). The documentation of [`sqlite3_column_value`](https://www.sqlite.org/c3ref/column_blob.html) seems to recommend against this design:

>**Warning:** The object returned by `sqlite3_column_value()` is an unprotected `sqlite3_value object`. In a multithreaded environment, an unprotected `sqlite3_value` object may only be used safely with `sqlite3_bind_value()` and `sqlite3_result_value()`. If the unprotected `sqlite3_value` object returned by `sqlite3_column_value()` is used in any other way, including calls to routines like `sqlite3_value_int()`, `sqlite3_value_text()`, or `sqlite3_value_bytes()`, the behavior is not threadsafe. Hence, the `sqlite3_column_value()` interface is normally only useful within the implementation of application-defined SQL functions or virtual tables, not within top-level application code.

Reply:
> Hence, the sqlite3_column_value() interface is normally only useful within the implementation of application-defined SQL functions

Isn't that exactly what we're doing? `unprotected` just means that there's no mutex around the value, but that literally just means that it needs to be `!Sync` (which it will be by default if it involves a pointer)

Reply:
>Isn't that exactly what we're doing?
Of course it is. The question I am trying to address is whether or not it is a good idea to change the preexisting `FromSql` code for handling the column values from normal database reads.

>`unprotected` just means that there's no mutex around the value
Ok. The way I read the documentation it seemed to me that the only safe way to access the value of an `unprotected` `sqlite3_value` would be via `sqlite3_value_dup`. I see now that the documentation is not as strict, it seems we can handle the thread safety ourselves. Good.

I'll look into refactoring to using `sqlite3_value`s for columns as well, so we can get rid of the duplication.